pmlc/dialect/pxa/transforms/cache.cc

```diff
@@ -251,10 +269,191 @@ Optional<CacheInfo> cacheReduce(AffineParallelOp par, PxaReduceOp reduce) {
+                           Value cache, const RelativeAccessPattern &rap) {
+  // Make a new reduce and remove the old one
+  auto loc = reduce.getLoc();
+  auto innerMap = convertToValueMap(band.getContext(), rap.inner);
+  OpBuilder builder(reduce);
+  auto newReduce = builder.create<PxaReduceOp>(loc, reduce.agg(), reduce.val(),
+                                               cache, innerMap.getAffineMap(),
+                                               innerMap.getOperands());
+  reduce.replaceAllUsesWith(newReduce.result());
+  reduce.erase();
+
+  // Walk upwards to adjust types
+  Value result = newReduce;
+  Type newType = newReduce.getType();
+  while (result.getParentBlock() != band.getBody()) {
+    auto use = result.use_begin();
+    result = getNextIndirectUse(*use);
+    // if (!result) {
+    //   band.emitOpError("No indirect uses found");
+    // }
+    if (auto ifOp = dyn_cast<AffineIfOp>(result.getDefiningOp())) {
+      // TODO: should we fail in this case?
+      auto yield =
+          dyn_cast<AffineYieldOp>(ifOp.getElseBlock()->getTerminator());
+      yield.setOperand(0, cache);
+    }
+    result.setType(newType);
+  }
+  return result;
+}
+
+void CachePlan::execute() {
+  for (auto &[memref, entry] : entries) {
+    // determine the level to cache at
+    entry.band = getRelevantBand(entry.rap, outerBand, middleBand);
+    auto loc = entry.band.getLoc();
+    auto builder = OpBuilder::atBlockBegin(entry.band.getBody());
+    SmallVector<StrideInfo, 4> zeroOffset(entry.rap.innerCount.size());
+
+    auto cache = allocateLocalCache(builder, memref, loc, entry.rap);
+    if (isInitialized(memref)) {
+      // copy global -> local
+      entry.copyInto = true;
+      auto copyLoop =
+          createCopyLoop(builder, loc, entry.rap.innerCount, memref, cache,
+                         entry.rap.outer, zeroOffset, AtomicRMWKind::assign);
+      copyLoop.getOperation()->setAttr("cache_in", builder.getUnitAttr());
+      cache = copyLoop.getResult(0);
+    }
+
+    for (const auto &load : entry.loads) {
+      replaceLoad(load.op, cache, load.rap);
+    }
+
+    if (entry.reduces.size()) {
+      Value finalValue;
+      for (const auto &reduce : entry.reduces) {
+        finalValue = replaceReduce(entry.band, reduce.op, cache, reduce.rap);
+      }
+
+      // copy local -> global
+      entry.copyFrom = true;
+      OpBuilder::InsertionGuard guard(builder);
+      auto yield = entry.band.getBody()->getTerminator();
+      builder.setInsertionPoint(yield);
+      auto &finalUse = *finalValue.use_begin();
+      auto copyLoop =
+          createCopyLoop(builder, loc, entry.rap.innerCount, finalValue, memref,
+                         zeroOffset, entry.rap.outer, AtomicRMWKind::assign);
+      copyLoop.getOperation()->setAttr("cache_out", builder.getUnitAttr());
+      finalUse.set(copyLoop.getResult(0));
+    }
+  }
+}
+
+struct CachePass : public CacheBase<CachePass> {
+  void runOnFunction() final {
```

Comment:
Some of this is a bit more use case specific than it needs to be, but we can refactor it when we need to.

Reply:
Yeah, and also this is meant for testing only at the moment.